File: examples/ch12/ts/bundle.ts
// Chapter 12 · Composite: treat a tree of line items uniformly.
//
// Product is a leaf; Bundle is a composite holding other line items (products
// or sub-bundles). Both satisfy LineItem, so a cart totals itself with no type
// checks and no special-casing of branches versus leaves.

export interface LineItem {
  price(): number;
}

export class Product implements LineItem {
  constructor(
    readonly name: string,
    private readonly amount: number,
  ) {}
  price(): number {
    return this.amount;
  }
}

export class Bundle implements LineItem {
  private readonly items: LineItem[];
  constructor(
    readonly name: string,
    items: LineItem[] = [],
  ) {
    this.items = [...items];
  }
  add(item: LineItem): void {
    this.items.push(item);
  }
  price(): number {
    return this.items.reduce((sum, item) => sum + item.price(), 0); // recurse
  }
}

export const cartTotal = (items: readonly LineItem[]): number =>
  items.reduce((sum, item) => sum + item.price(), 0); // leaf and branch alike

File: examples/ch12/py/bundle.py
"""Chapter 12 · Composite: treat a tree of line items uniformly.

A Product is a leaf; a Bundle is a composite that holds other line items —
products or sub-bundles. Both answer price(), so a cart totals itself with no
isinstance checks and no special-casing of branches versus leaves.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class LineItem(ABC):
    """The component interface — shared by leaf and composite alike."""

    @abstractmethod
    def price(self) -> float: ...


class Product(LineItem):
    """A leaf: a single sellable item."""

    def __init__(self, name: str, price: float) -> None:
        self.name = name
        self._price = price

    def price(self) -> float:
        return self._price


class Bundle(LineItem):
    """A composite: a sellable thing made of other line items."""

    def __init__(self, name: str, items: list[LineItem] | None = None) -> None:
        self.name = name
        self._items: list[LineItem] = list(items) if items else []

    def add(self, item: LineItem) -> None:
        self._items.append(item)

    def price(self) -> float:
        return sum(item.price() for item in self._items)  # recurse, uniformly


def cart_total(items: list[LineItem]) -> float:
    """The client treats leaf and branch identically — no type checks."""
    return sum(item.price() for item in items)
